File: src/main/java/de/jonihoffi/plugins/enhavo/NewBlockYamlFileAction.java
package de.jonihoffi.plugins.enhavo;

import com.intellij.ide.IdeView;
import com.intellij.ide.fileTemplates.FileTemplateManager;
import com.intellij.ide.fileTemplates.FileTemplateUtil;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiFile;
import de.jonihoffi.plugins.enhavo.dialog.NewBlockYamlFileDialog;
import org.jetbrains.annotations.NotNull;

import java.util.Properties;

public class NewBlockYamlFileAction extends AnAction {
    public NewBlockYamlFileAction() {
        super("New Enhavo Block", "Creates a new enhavo-block yaml file", null);
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent event) {
        Project project = event.getProject();
        if (project == null) return;

        IdeView view = event.getData(LangDataKeys.IDE_VIEW);
        if (view == null) return;

        PsiDirectory directory = view.getOrChooseDirectory();
        if (directory == null) return;

        NewBlockYamlFileDialog dialog = new NewBlockYamlFileDialog();
        if (dialog.showAndGet()) {
            String name = dialog.getName();
            if (!name.isEmpty()) {
                String nameWithoutExtension = name.endsWith(".yaml") ? name.substring(0, name.length() - 5) : name;
                createBlockYamlFile(project, directory, nameWithoutExtension, dialog);
            } else {
                Messages.showErrorDialog(project, "The file name cannot be empty.", "Error");
            }
        }
    }

    @Override
    public void update(@NotNull AnActionEvent event) {
        IdeView view = event.getData(LangDataKeys.IDE_VIEW);
        if (view == null) return;

        PsiDirectory[] directories = view.getDirectories();
        boolean enabled = false;
        for (PsiDirectory dir : directories) {
            if (isBlockDirectory(dir.getVirtualFile())) {
                enabled = true;
                break;
            }
        }

        event.getPresentation().setEnabledAndVisible(enabled);
    }

    @Override
    public @NotNull ActionUpdateThread getActionUpdateThread() {
        return ActionUpdateThread.BGT;
    }

    private boolean isBlockDirectory(VirtualFile file) {
        return file.isDirectory() && ("block".equals(file.getName()) || "block".equals(file.getParent().getName()));
    }

    private void createBlockYamlFile(Project project, PsiDirectory directory, String name, NewBlockYamlFileDialog dialog) {
        try {
            String templateName = dialog.includeItems() ? "Block Item YAML File Template" : "Block YAML File Template";
            Properties properties = FileTemplateManager.getInstance(project).getDefaultProperties();
            properties.setProperty("NAME", name);
            properties.setProperty("CAMELCASE", lowerCaseFirstChar(name + "Block"));
            properties.setProperty("GROUPS", dialog.getGroups());

            var template = FileTemplateManager.getInstance(project).getInternalTemplate(templateName);
            PsiFile file = FileTemplateUtil.createFromTemplate(template, name + "Block.yaml", properties, directory).getContainingFile();
            openFileInEditor(project, file.getVirtualFile());
        } catch (Exception e) {
            Messages.showErrorDialog(project, "Error creating file: " + e.getMessage(), "Error");
        }
    }

    private void openFileInEditor(Project project, VirtualFile virtualFile) {
        FileEditorManager.getInstance(project).openFile(virtualFile, true);
    }

    private String lowerCaseFirstChar(String input) {
        if (input == null || input.isEmpty()) {
            return input;
        }
        char firstChar = Character.toLowerCase(input.charAt(0));
        if (input.length() == 1) {
            return String.valueOf(firstChar);
        }
        return firstChar + input.substring(1);
    }
}

File: src/main/kotlin/de/jonihoffi/plugins/demo/NewBlockYamlFileAction.kt
package de.jonihoffi.plugins.demo

import com.intellij.ide.fileTemplates.FileTemplateManager
import com.intellij.ide.fileTemplates.FileTemplateUtil
import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiFile
import com.intellij.ui.components.dialog
import de.jonihoffi.plugins.demo.dialog.NewBlockYamlFileDialog

class NewBlockYamlFileAction : AnAction("New enhavo-block yaml file", "Creates a new enhavo-block yaml file", null) {

    override fun actionPerformed(event: AnActionEvent) {
        val project = event.project ?: return
        val view = event.getData(LangDataKeys.IDE_VIEW) ?: return
        val directory = view.orChooseDirectory ?: return

        val dialog = NewBlockYamlFileDialog()
        if (dialog.showAndGet()) {
            val name = dialog.getName()
            if (name.isNotEmpty()) {
                val nameWithoutExtension = if (name.endsWith(".yaml")) name.removeSuffix(".yaml") else name
                createBlockYamlFile(project, directory, nameWithoutExtension, dialog)
            } else {
                Messages.showErrorDialog(project, "The file name cannot be empty.", "Error")
            }
        }
    }

    override fun update(event: AnActionEvent) {
        val project = event.project
        val view = event.getData(LangDataKeys.IDE_VIEW)
        val directories = view?.directories ?: return

        event.presentation.isEnabledAndVisible = directories.any { dir ->
            isBlockDirectory(dir.virtualFile)
        }
    }

    override fun getActionUpdateThread(): ActionUpdateThread {
        return ActionUpdateThread.BGT
    }

    private fun isBlockDirectory(file: VirtualFile): Boolean {
        return file.isDirectory && (file.name == "block" || file.parent?.name == "block")
    }

    private fun createBlockYamlFile(project: Project, directory: PsiDirectory, name: String, dialog: NewBlockYamlFileDialog) {
        try {
            val templateName = if (dialog.includeItems()) "Block Item YAML File Template" else "Block YAML File Template"
            val properties = FileTemplateManager.getInstance(project).defaultProperties
            properties["NAME"] = name
            properties["CAMELCASE"] = lowerCaseFirstChar(name + "Block")
            properties["GROUPS"] = dialog.getGroups()

            val template = FileTemplateManager.getInstance(project).getInternalTemplate(templateName)
            val file: PsiFile = FileTemplateUtil.createFromTemplate(template, name + "Block.yaml", properties, directory) as PsiFile
            openFileInEditor(project, file.virtualFile)
        } catch (e: Exception) {
            Messages.showErrorDialog(project, "Error creating file: ${e.message}", "Error")
        }
    }

    private fun openFileInEditor(project: Project, virtualFile: VirtualFile) {
        FileEditorManager.getInstance(project).openFile(virtualFile, true)
    }

    private fun lowerCaseFirstChar(input: String?): String? {
        if (input.isNullOrEmpty()) {
            return input
        }
        val firstChar = input[0].lowercaseChar()
        if (input.length == 1) {
            return firstChar.toString()
        }
        return firstChar.toString() + input.substring(1)
    }
}